subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java

```diff
@@ -102,11 +105,68 @@
  */
 @DisableCachingByDefault(because = "Abstract super-class, not to be instantiated directly")
 public abstract class AbstractTestTask extends ConventionTask implements VerificationTask, Reporting<TestTaskReports> {
+
+    /**
+     * Wraps a list of listeners to subscribe, and lazily configures an anonymous broadcaster with those listeners when requested.
+     * Instances of this class are suitable for serialization (as long as listeners are serializable as well).
+     */
+    private class BroadcastSubscriptions<T> {
+        private final Class<T> listenerClass;
+        private final List<Object> subscribedListeners = new LinkedList<Object>();
+        private transient ListenerBroadcast<T> broadcaster;
+
+        private BroadcastSubscriptions(Class<T> listenerClass) {
+            this.listenerClass = listenerClass;
+        }
+
+        @SuppressWarnings("unchecked")
+        public synchronized ListenerBroadcast<T> get() {
+            if (broadcaster == null) {
+                broadcaster = getListenerManager().createAnonymousBroadcaster(listenerClass);
+                for (Object listener: subscribedListeners) {
+                    if (listenerClass.isInstance(listener)) {
+                        broadcaster.add((T) listener);
+                    } else {
+                        broadcaster.add((Dispatch<MethodInvocation>) listener);
+                    }
+                }
+            }
+            return broadcaster;
+        }
+
+        public synchronized void addListener(T listener) {
```

Comment:
❓ `synchronized`?

Reply:
What is the concern here with using `synchronized`, is it because it is not needed as there is no possibility of parallelism? My motivation was the laziness of the broadcaster. With the mutex, we guarantee a listener that signed up for a broadcaster is always added to the broadcaster, instead of risking being ignored because it coincided with the broadcaster being lazily created and populated.

Reply:
Well, without configuration cache / project isolation in the picture I guess it would be possible to come up with a contrived way to execute these methods concurrently but notice the original implementation based on `ListenerBroadcast` was *not* thread-safe.

Reply:
Unless we have strong guarantees this code cannot be executed from multiple threads, I would rather not add another instance of lazy init'ing code that is not thread-safe. Investigating and fixing concurrency issues is really hard, I would not like to add technical debt unless we have a good reason to do so.

Reply:
I don't think the Gradle vintage execution model provides strong guarantees in that direction BUT the norm is that a task CANNOT be configured from multiple threads in parallel.  

Configuration cache does provide stronger guarantees since it isolates tasks from each other.

Reply:
If we prevent adding listeners at execution time, then the configuration time guarantee will be enough. Thanks!

Reply:
Removed lazy init'ing and synchronization